networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/Eth2PeerSelectionStrategy.java

```diff
@@ -38,54 +47,149 @@
   private final TargetPeerRange targetPeerCountRange;
   private final PeerSubnetSubscriptions.Factory peerSubnetSubscriptionsFactory;
   private final ReputationManager reputationManager;
+  private final Shuffler shuffler;
 
   public Eth2PeerSelectionStrategy(
       final TargetPeerRange targetPeerCountRange,
       final PeerSubnetSubscriptions.Factory peerSubnetSubscriptionsFactory,
-      final ReputationManager reputationManager) {
+      final ReputationManager reputationManager,
+      final Shuffler shuffler) {
     this.targetPeerCountRange = targetPeerCountRange;
     this.peerSubnetSubscriptionsFactory = peerSubnetSubscriptionsFactory;
     this.reputationManager = reputationManager;
+    this.shuffler = shuffler;
   }
 
   @Override
   public List<PeerAddress> selectPeersToConnect(
-      final P2PNetwork<?> network, final Supplier<List<DiscoveryPeer>> candidates) {
+      final P2PNetwork<?> network,
+      final PeerSources peerSources,
+      final Supplier<List<DiscoveryPeer>> candidates) {
     final PeerSubnetSubscriptions peerSubnetSubscriptions =
         peerSubnetSubscriptionsFactory.create(network);
     final int peersRequiredForPeerCount =
         targetPeerCountRange.getPeersToAdd(network.getPeerCount());
+    final int randomlySelectedPeerCount = getCurrentRandomlySelectedPeerCount(network, peerSources);
+    final int randomlySelectedPeersToAdd =
+        targetPeerCountRange.getRandomlySelectedPeersToAdd(randomlySelectedPeerCount);
+
     final int peersRequiredForSubnets = peerSubnetSubscriptions.getSubscribersRequired();
-    final int maxAttempts = Math.max(peersRequiredForPeerCount, peersRequiredForSubnets);
-    LOG.trace("Connecting to up to {} known peers", maxAttempts);
-    if (maxAttempts == 0) {
+    final int scoreBasedPeersToAdd =
+        Math.max(peersRequiredForPeerCount - randomlySelectedPeerCount, peersRequiredForSubnets);
+    LOG.trace(
+        "Connecting to up to {} known peers", scoreBasedPeersToAdd + randomlySelectedPeersToAdd);
+    if (scoreBasedPeersToAdd == 0 && randomlySelectedPeersToAdd == 0) {
       return emptyList();
     }
```

Comment:
I'm in two minds - it requires pulling in every discovered peer which isn't the end of the world but definitely isn't free. Given we also get to avoid the cost of calculating peer scores etc I'm inclined to keep bailing out early.